[Misc][Quark] Upstream Quark format to VLLM

kewang-amd opened the pull request:
This PR is related to the issue discussed in this https://github.com/vllm-project/vllm/issues/10294.

Commit:
[AMD] support Quark quantized format

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
[Model] add param_name remap in experts in dbrx model

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
[Model] support quark kv cache format in dbrx

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
add kv cache remap method for quark format

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
[AMD][Quark] Fix fails in pr checks

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

kewang-amd committed:
[AMD][Quark] replace get_compressed_tensors_cache_scale with get_cache_scale in aria.py

Signed-off-by: kewang2 <kewang2@amd.com>

Commit:
[AMD][Quark] remove quark dependency

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
[AMD][Quark] fix mypy error

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
auto fix yapf

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
fix mypy error

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
fix code format

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

robertgshaw2-neuralmagic commented:
Its a bit unfortunate that:
- We will have all this duplicated code
- We will have multiple model formats to maintain

However, I understand that AMD wants to push forward with Quark, so I think this approach is very reasonable. I would like to take on this effort myself and do two things:
- Factor out some of the utility functions such that we do not have too much copy / paste code between the backends
- Leverage the `Kernel` abstraction such that `compressed-tensors` models can run on RoCM backend

To do this, it would be very helpful if the Quark team could provide:
- A sample `W8A8-int8` model
- A sample `W8A8-fp8` model

If this is done, I can take on the effort of finalizing the integration. Thanks!

kewang-xlnx commented:
> Its a bit unfortunate that:
> 
> * We will have all this duplicated code
> * We will have multiple model formats to maintain
> 
> However, I understand that AMD wants to push forward with Quark, so I think this approach is very reasonable. I would like to take on this effort myself and do two things:
> 
> * Factor out some of the utility functions such that we do not have too much copy / paste code between the backends
> * Leverage the `Kernel` abstraction such that `compressed-tensors` models can run on RoCM backend
> 
> To do this, it would be very helpful if the Quark team could provide:
> 
> * A sample `W8A8-int8` model
> * A sample `W8A8-fp8` model
> 
> If this is done, I can take on the effort of finalizing the integration. Thanks!

Hi @robertgshaw2-neuralmagic. Thank you for your comments. 

Your proposed work is highly valuable and will benefit both VLLM contributors and users. Factoring out utility functions to reduce code duplication and leveraging the Kernel abstraction for compatibility with the RoCM backend are excellent steps forward.

Here is a config file of W8A8-fp8: 
[config.json](https://github.com/user-attachments/files/18145163/config.json)
And this is a config file of W8A8-int8:
[config.json](https://github.com/user-attachments/files/18145165/config.json)

key names and data types of Quark safetensors
```
model.layers.1.self_attn.k_proj.input_scale, 	torch.float16
model.layers.1.self_attn.k_proj.weight, 	torch.float8_e4m3fn
model.layers.1.self_attn.k_proj.weight_scale, 	torch.float16
model.layers.1.self_attn.o_proj.input_scale, 	torch.float16
model.layers.1.self_attn.o_proj.weight, 	torch.float8_e4m3fn
model.layers.1.self_attn.o_proj.weight_scale, 	torch.float16
model.layers.1.self_attn.q_proj.input_scale, 	torch.float16
model.layers.1.self_attn.q_proj.weight, 	torch.float8_e4m3fn
model.layers.1.self_attn.q_proj.weight_scale, 	torch.float16
model.layers.1.self_attn.v_proj.input_scale, 	torch.float16
model.layers.1.self_attn.v_proj.weight, 	torch.float8_e4m3fn
model.layers.1.self_attn.v_proj.weight_scale, 	torch.float16
```
KV scale format if kv cache used
```
model.layers.1.self_attn.k_proj.output_scale, 	torch.float16
model.layers.1.self_attn.v_proj.output_scale, 	torch.float16
```

And it’s important that the integration of the Quark format into VLLM is a priority. To ensure long-term maintainability, we would like to keep the responsibility for maintaining the Quark format within AMD’s team. Our team will closely monitor the integration to minimize any maintenance overhead for the vLLM project.

Let us know if you need any further support during the process. Thanks again for your collaboration!

mgoin commented:
> KV scale format if kv cache used
> ```
> model.layers.1.self_attn.k_proj.output_scale, 	torch.float16
> model.layers.1.self_attn.v_proj.output_scale, 	torch.float16
> ```
Hi @kewang-xlnx can you describe what happens in the case of a model with fused QKV proj, such as Phi-3? In several cases models do not have separate k_proj/v_proj linear layers with which to register output scales on. This is why we [register](https://github.com/neuralmagic/compressed-tensors/blob/5771decf62d8494524246f06836a19e53e6d5612/src/compressed_tensors/quantization/lifecycle/initialize.py#L217) explicit `self_attn.k_scale` and `self_attn.v_scale` in compressed-tensors

mgoin commented on vllm/model_executor/layers/quantization/quark/quark.py:
```diff
@@ -0,0 +1,386 @@
+        if name.endswith(".output_scale"):
+            if len(kv_proj_names) == 1 and kv_proj_names[0] in name:
+                kv_output_scale_name = "." + kv_proj_names[0] + ".output_scale"
+                return [
+                    name.replace(kv_output_scale_name, ".attn.k_scale"),
+                    name.replace(kv_output_scale_name, ".attn.v_scale")
+                ]
```
What is the purpose of returning multiple output names for the same input name? I think if you just return a string like before, you won't need to change all of the call-sites aside from
```
            if (self.quant_config is not None and
                (scale_name := self.quant_config.get_cache_scale(name))):
```

kewang-xlnx replied:
Are you suggesting that the k_proj and v_proj here are represented by only one layer, and we can remap the corresponding kv_scale only to the attn.k_scale in the vllm, and then the v_scale can also be assigned through this line of code https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/kv_cache.py#L56 ? If that's the case, indeed, a one-to-one remapping is sufficient, but we must ensure that this code segment (https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/kv_cache.py#L56) is not altered.

mgoin replied:
Yes, you found the right place. This is an assumption I expect we will keep and would prefer we keep the duplication internal to the layer initialization, rather than augmenting the weight loading as you've done here. I would appreciate if you made the change

kewang-xlnx replied:
ok, thanks. I have made the change.

mgoin commented on vllm/model_executor/models/aria.py:
```diff
@@ -390,14 +388,18 @@ def load_weights(self, weights: Iterable[Tuple[str,
-                loaded_weight = loaded_weight[0]
-                weight_loader(param, loaded_weight)
-                loaded_params.add(scale_name)
+            if (self.quant_config is not None and
+                (scale_names := self.quant_config.get_cache_scale(name))):
+                # Loading kv cache scales for quark and
+                # compressed-tensors quantization
+                for scale_name in scale_names:
+                    param = params_dict[scale_name]
+                    weight_loader = getattr(param, "weight_loader",
+                                            default_weight_loader)
+                    loaded_weight = (loaded_weight if loaded_weight.dim() == 0
+                                     else loaded_weight[0])
+                    weight_loader(param, loaded_weight)
+                    loaded_params.add(scale_name)
```
Like I said before, why does `get_cache_scale` need to return a list? This would be simpler if it was a 1-to-1 function.

Also what is the purpose of the loaded_weight condition? I actually think it would be better to treat it as a tensor in the long run rather than a scalar, so we can move beyond per-tensor in the future. This is out of scope for this PR though

kewang-xlnx replied:
In quark's model, if it is per_tensor quantization, the corresponding scale is a scalar expressed as torch.tensor(scale). Using tensor[0] to obtain it will report an error, so we added a judgment. This will also increase the robustness of the system to prevent scale from being a scale.

kewang-xlnx replied:
Okay, we will consider better ways to solve this problem in the future.

mgoin commented on vllm/model_executor/models/dbrx.py:
```diff
@@ -83,33 +83,45 @@ def __init__(
-                [-1, self.intermediate_size * self.tp_size, self.d_model],
-            )
-            param_data[:, 0:shard_size, :] = loaded_weight[:, shard, :]
+            if param_name.endswith("weight"):
+                loaded_weight = torch.reshape(
+                    loaded_weight,
+                    [-1, self.intermediate_size * self.tp_size, self.d_model],
+                )
+                param_data[:, 0:shard_size, :] = loaded_weight[:, shard, :]
+            elif param_name.endswith("weight_scale"):
+                param_data[:, 0] = loaded_weight
+            else:
+                param_data = loaded_weight
```
Why do we need to change the weight loader?

kewang-xlnx replied:
The dbrx module of vllm only considered the load of weight before, not the load of scale. Here, the load of scale is added. In fact, I copied this code from rocm/vllm.

Commit:
Update vllm/model_executor/layers/quantization/compressed_tensors/triton_scaled_mm.py

Great suggestion to use reshape operator. I think the change should be like this
scale_a = scale_a.reshape(-1, 1) if scale_a.dim() <= 1 else scale_a
scale_b = scale_b.reshape(-1, 1) if scale_b.dim() <= 1 else scale_b

Co-authored-by: Michael Goin <michael@neuralmagic.com>
Signed-off-by: kewang-xlnx <kewang@xilinx.com>

kewang-xlnx commented on vllm/model_executor/layers/quantization/quark/quark.py:
```diff
@@ -0,0 +1,386 @@
+        if len(kv_cache_group) == 0:
+            kv_cache_config = None
+        else:
+            kv_cache_set = set(kv_cache_group)
```
Sure, i will add some comments on this part of codes.

kewang-xlnx commented on vllm/model_executor/layers/quantization/quark/quark.py:
```diff
@@ -0,0 +1,386 @@
+    def get_cache_scale(self, name: str) -> Optional[List[str]]:
+        """
+        Check whether the param name matches the format for k/v cache scales
+        in compressed-tensors. If this is the case, return its equivalent
```
Yes, thanks.

kewang-xlnx commented:
> > KV scale format if kv cache used
> > ```
> > model.layers.1.self_attn.k_proj.output_scale, 	torch.float16
> > model.layers.1.self_attn.v_proj.output_scale, 	torch.float16
> > ```
> 
> Hi @kewang-xlnx can you describe what happens in the case of a model with fused QKV proj, such as Phi-3? In several cases models do not have separate k_proj/v_proj linear layers with which to register output scales on. This is why we [register](https://github.com/neuralmagic/compressed-tensors/blob/5771decf62d8494524246f06836a19e53e6d5612/src/compressed_tensors/quantization/lifecycle/initialize.py#L217) explicit `self_attn.k_scale` and `self_attn.v_scale` in compressed-tensors. We could likely reuse the same utility if you choose the same place to register these (on the Attention module)

Hi @mgoin. If the kv_scale of fused  QKV proj in quark is as below, take Phi-3 for example.
`model.layers.22.self_attn.qkv_proj.output_scale`
This is why we need to remap one kv_scale to both attn.k_scale and attn.v_scale.

kewang-xlnx commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:
```diff
@@ -332,6 +332,26 @@ def get_scheme(
 
         return scheme
 
+    # move the get_compressed_tensors_cache_scale method from
+    # utils.py to instance method of CompressedTensorsConfig
+    # class. By doing this, different QuantizationConfig
+    # classes can implement their own get_cache_scale method.
```
I don’t quite understand the meaning of this comment

kewang-xlnx replied:
I see, I will delete this comments.

Commit:
Modify based on PR comments

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
fix mypy error

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
delete comments of get_cache_scale in compressed_tensors.py

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

mgoin reviewed (approved):
Apologies for the delay and thanks for the changes! This seems to be in a decent place to land. Would you have a small quark dense and/or moe model we could add to CI @kewang-xlnx ?

Commit:
fix mypy error

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
change get_compressed_tensors_cache_scale to get_cache_scale in gemma2.py

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
add CI test for quark format

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
fix mypy error in CI tests

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

Commit:
be consistent with origin vllm

Signed-off-by: kewang-xlnx <kewang@xilinx.com>

mgoin commented:
Thank you for your hard work @kewang-xlnx!
